File: YouChatServer/src/chat/util/TransferUtil.java
package chat.util;

import chat.ChatRunner;
import chat.msg.Message;

import java.io.IOException;
import java.io.ObjectOutputStream;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.util.Date;

public class TransferUtil {
    //server
    private static final int serverReceivePort = 8080;
    private static final int serverSendPort = 8081;

    private static DatagramSocket serverSendSocket;
    private static DatagramSocket serverReceiveSocket;

    static {
        try {
            serverReceiveSocket = new DatagramSocket(serverReceivePort);
            serverSendSocket = new DatagramSocket(serverSendPort);
        } catch (SocketException e) {
            e.printStackTrace();
        }
    }

    private TransferUtil() {
    }

    public static boolean sendByUDP(Message message) {
        try {
            byte[] bytes = SerializeUtil.object2Bytes(message);
            if (bytes == null) {
                System.out.println("sendByUDP: message为空，使用UDP发送失败");
                return false;
            }
            if (bytes.length > 1024) {
                System.out.println("sendByUDP: message长度大于1024，使用UDP发送失败");
                return false;
            }
            // todo 根据to的账号获取ip
            InetAddress toAddr = InetAddress.getByName(ChatRunner.idIPMap.get(message.getTo()).ip);
            DatagramPacket packet = new DatagramPacket(bytes, bytes.length, toAddr, ChatRunner.idIPMap.get(message.getTo()).udpPort);

            // 发送消息到服务器,阻塞接受服务器的回复
            byte[] bytesReceived = new byte[1024];
            DatagramPacket packetReceived = new DatagramPacket(bytesReceived, bytesReceived.length);

            serverSendSocket.send(packet);
            serverSendSocket.receive(packetReceived);

            Message receiveMsg = (Message) SerializeUtil.bytes2Object(packetReceived.getData());
            // todo 考虑response的内容
            assert receiveMsg != null;
            return new String(receiveMsg.getData()).equals("OK");
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    public static Message receiveByUDP() {
        Message response = new Message("server", "", null, new Date(), Message.Type.TEXT);
        try {
            // 要接收的报文
            byte[] bytes = new byte[1024];
            DatagramPacket packet = new DatagramPacket(bytes, bytes.length);

            // 接收socket客户端发送的数据。如果未收到会一直阻塞
            serverReceiveSocket.receive(packet);
            System.out.println("from: addr:" + packet.getAddress() + "\tport: " + packet.getPort());

            Message message = (Message) SerializeUtil.bytes2Object(packet.getData());
            assert message != null;

            if (Message.Type.UDP_PORT_INFO == message.getDataType()) {
                //todo 安全性没有保证
                if(ChatRunner.idIPMap.get(message.getFrom())==null){
                    ChatRunner.idIPMap.put(message.getFrom(), new ChatRunner.PortInfo(packet.getAddress().getHostAddress(),packet.getPort(),-1));
                }
                else{
                    ChatRunner.idIPMap.get(message.getFrom()).ip = packet.getAddress().getHostAddress();
                    ChatRunner.idIPMap.get(message.getFrom()).udpPort = packet.getPort();
                }
            } else if (Message.Type.TCP_PORT_INFO == message.getDataType()) {
                //todo 安全性没有保证
                if(ChatRunner.idIPMap.get(message.getFrom())==null){
                    ChatRunner.idIPMap.put(message.getFrom(), new ChatRunner.PortInfo(packet.getAddress().getHostAddress(),-1,packet.getPort()));
                }
                else{
                    ChatRunner.idIPMap.get(message.getFrom()).ip = packet.getAddress().getHostAddress();
                    ChatRunner.idIPMap.get(message.getFrom()).tcpPort = packet.getPort();
                }
            }

            response.setData("OK".getBytes(StandardCharsets.UTF_8));
            response.setTo(message.getFrom());
            byte[] respBytes = SerializeUtil.object2Bytes(response);
            assert respBytes != null;
            serverReceiveSocket.send(new DatagramPacket(respBytes, respBytes.length, packet.getAddress(), packet.getPort()));


//            System.out.println(packet.getLength());
            return message;
        } catch (IOException e) {
            e.printStackTrace();
        }

        //todo return;
        return null;
    }

    public static void sendByTCP(Socket socket, Message msg) {
        try {
            ObjectOutputStream oos = new ObjectOutputStream(socket.getOutputStream());
            oos.writeObject(msg);
            oos.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

}

File: app/src/main/java/edu/buptsse/youchat/util/communication/TransferUtil.kt
package edu.buptsse.youchat.util.communication

import android.util.Log
import edu.buptsse.youchat.Message
import edu.buptsse.youchat.chat.CallActivity
import edu.buptsse.youchat.main.curUser
import edu.buptsse.youchat.main.msgMap
import edu.buptsse.youchat.service.CommunicationService
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.EOFException
import java.io.IOException
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.net.*
import java.nio.charset.StandardCharsets
import java.util.*


private const val serverReceivePort = 8080
private const val serverSendPort = 8081
private const val tcpChatServerPort = 10086
private const val tcpCallServerPort = 37040

//client
private const val clientSendPort = 8088 //负责向服务端发送消息并接受回复
private const val clientReceivePort = 8089 //接受服务器主动发送的消息

const val CALL_REQUEST = "req"
const val CALL_ACCEPT = "acc"

//    private static final int POOL_SIZE = Runtime.getRuntime().availableProcessors() * 2;
//    private static final ThreadPoolExecutor executors = new ThreadPoolExecutor(POOL_SIZE, POOL_SIZE, 0, TimeUnit.SECONDS, new ArrayBlockingQueue<>(128));
//server
private var serverAddr: InetAddress? = null

//socket
private var clientSendSocket: DatagramSocket? = null //负责向服务端发送消息并接受回复
private var clientReceiveSocket: DatagramSocket? = null //接受服务器主动发送的消息
private var chatTcpSocket: Socket? = null
var callTcpSocket: Socket? = null

suspend fun transferInit() {
    withContext(Dispatchers.IO) {
        try {
            serverAddr = InetAddress.getByName("119.3.224.100")
            //            serverAddr = InetAddress.getByName("127.0.0.1");
            //clientReceiveSocket = DatagramSocket(clientReceivePort)
            //clientSendSocket = DatagramSocket(clientSendPort)
            chatTcpSocket = Socket(serverAddr, tcpChatServerPort)
            PortInfoUtil.sendPortInfo(curUser.id)
            callTcpSocket = Socket(serverAddr, tcpCallServerPort)
            val oos1 = ObjectOutputStream(callTcpSocket!!.getOutputStream())
            oos1.writeObject(Message(curUser.id, curUser.id, null, Date(), Message.Type.CALL))
            oos1.flush()
            Log.e("socket", callTcpSocket.toString())
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}

suspend fun connectCallServer() {
    withContext(Dispatchers.IO) {
        try {
            if (callTcpSocket!!.isClosed) {
                callTcpSocket!!.connect(InetSocketAddress(serverAddr, tcpCallServerPort))
                val oos1 = ObjectOutputStream(callTcpSocket!!.getOutputStream())
                oos1.writeObject(Message(curUser.id, curUser.id, null, Date(), Message.Type.CALL))
                oos1.flush()
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }
}

suspend fun sendByUDP(message: Message?): Boolean {
    withContext(Dispatchers.IO) {
        try {
            val bytes = SerializeUtil.object2Bytes(message)
            if (bytes == null) {
                println("sendByUDP: message为空，使用UDP发送失败")
                return@withContext false
            }
            if (bytes.size > 1024) {
                println("sendByUDP: message长度大于1024，使用UDP发送失败")
                return@withContext false
            }
            val packet = DatagramPacket(bytes, bytes.size, serverAddr, serverReceivePort)

            // 发送消息到服务器,阻塞接受服务器的回复
            val bytesReceived = ByteArray(1024)
            val packetReceived = DatagramPacket(bytesReceived, bytesReceived.size)
            clientSendSocket!!.send(packet)
            clientSendSocket!!.receive(packetReceived)
            val receiveMsg = (SerializeUtil.bytes2Object(packetReceived.data) as Message)
            return@withContext String(receiveMsg.data) == "OK"
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
    return false
}

suspend fun receiveByUDP(): Message? {
    withContext(Dispatchers.IO) {
        val response = Message("", "server", null, Date(), Message.Type.TEXT)
        try {
            // 要接收的报文
            val bytes = ByteArray(1024)
            val packet = DatagramPacket(bytes, bytes.size)

            // 接收socket客户端发送的数据。如果未收到会一直阻塞
            clientReceiveSocket!!.receive(packet)
            response.data = "OK".toByteArray(StandardCharsets.UTF_8)
            val respBytes = SerializeUtil.object2Bytes(response)!!
            clientReceiveSocket!!.send(DatagramPacket(respBytes, respBytes.size, serverAddr, serverSendPort))
            val message = SerializeUtil.bytes2Object(packet.data) as Message
            println(packet.length)
            return@withContext message
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    //todo return;
    return null
}

suspend fun sendChatByTCP(message: Message?) {
    withContext(Dispatchers.IO) {
        try {
            if (chatTcpSocket!!.isClosed) {
                chatTcpSocket!!.connect(InetSocketAddress(serverAddr, tcpChatServerPort))
                val oos1 = ObjectOutputStream(callTcpSocket!!.getOutputStream())
                oos1.writeObject(Message(curUser.id, curUser.id, null, Date(), Message.Type.CALL))
                oos1.flush()
            }
            val oos = ObjectOutputStream(chatTcpSocket!!.getOutputStream())
            oos.writeObject(message)
            oos.flush()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}

suspend fun sendCallReq(to: String, info: String) {
    withContext(Dispatchers.IO) {
        try {
            if (callTcpSocket!!.isClosed) {
                callTcpSocket!!.connect(InetSocketAddress(serverAddr, tcpCallServerPort))
                val oos1 = ObjectOutputStream(callTcpSocket!!.getOutputStream())
                oos1.writeObject(Message(curUser.id, curUser.id, null, Date(), Message.Type.CALL))
                oos1.flush()
            }
            val oos = ObjectOutputStream(callTcpSocket!!.getOutputStream())
            oos.writeObject(Message(curUser.id, to, info.toByteArray(), Date(), Message.Type.CALL))
            oos.flush()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}

suspend fun receiveCall(service: CommunicationService) {
    withContext(Dispatchers.IO) {
        try {
            if (callTcpSocket!!.isClosed) {
                callTcpSocket!!.connect(InetSocketAddress(serverAddr, tcpCallServerPort))
                val oos1 = ObjectOutputStream(callTcpSocket!!.getOutputStream())
                oos1.writeObject(Message(curUser.id, curUser.id, null, Date(), Message.Type.CALL))
                oos1.flush()
            }
            while (true) {
                try {
                    val ois = ObjectInputStream(callTcpSocket!!.getInputStream())
                    val msg = ois.readObject() as Message
                    Log.e("call", "${msg.from}: ${msg.text}")
                    if (msg.dataType == Message.Type.CALL) {
                        when (val response = msg.text) {
                            "acc" -> {
                                if (callTcpSocket!!.isClosed) {
                                    callTcpSocket!!.connect(InetSocketAddress(serverAddr, tcpCallServerPort))
                                }
                                CallActivity.isPickUp.value = true
                                Log.e("call", "accepted")
                                launch {
                                    record(callTcpSocket!!)
                                }
                                launch {
                                    play(callTcpSocket!!)
                                }
                                //                            new Thread(() -> AudioUtil.recordToSocket(callTcpSocket)).start();
                                //                            AudioUtil.play(callTcpSocket);
                            }

                            "req" -> {
                                Log.e("req", "收到请求")
                                service.startCallActivity(msg.from)
                                //                            new Thread(() -> AudioUtil.recordToSocket(callTcpSocket)).start();
                                //                            AudioUtil.play(callTcpSocket);
                            }

                            else -> {
                                println(response)
                            }
                        }
                    }
                } catch (e: EOFException) {
                    //不做处理，进行下一次循环，直到出现新的传入
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        }
    }
}

suspend fun receiveChatByTCP() {
    withContext(Dispatchers.IO) {
        try {
            if (chatTcpSocket!!.isClosed) {
                chatTcpSocket!!.connect(InetSocketAddress(serverAddr, tcpChatServerPort))
            }
            while (true) {
                try {
                    val ois = ObjectInputStream(chatTcpSocket!!.getInputStream())
                    val msg = ois.readObject() as Message
                    val list = msgMap[msg.from]
                    if (list == null) Log.e("null", "list")
                    list?.let {
                        it.add(Pair(false, msg))
                        msgMap[msg.from] = it
                    }
                    Log.e("msg", "${msg.from} ${msg.text}")
                } catch (e: EOFException) {
                    //不做处理，进行下一次循环，直到出现新的传入
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        }
    }
}

suspend fun record() {
    record(callTcpSocket!!)
}

suspend fun play() {
    play(callTcpSocket!!)
}
